### PR TITLE
[add] Add nmf in jax

### LINKED ISSUE #260: 使用 SPU 实现非负矩阵分解算法(NMF)
> 此 ISSUE 为 [隐语开源共建计划（SecretFlow Open Source Contribution Plan，简称 SF OSCP）第二期](https://github.com/orgs/secretflow/discussions/741)任务 ISSUE，欢迎社区开发者参与共建～
> 若有感兴趣想要认领的任务，但还未报名，辛苦先完成[报名](https://www.wjx.top/vm/QKio0dq.aspx# )进行哈～

## 任务介绍

- 任务名称：使用 SPU 实现非负矩阵分解算法(NMF)
- 技术方向：SPU/SML
- 任务难度：挑战🌟🌟🌟

## 详细要求
- 安全性（尽量少 reveal）
- 功能性：实现算法的基本功能，包括：
  - 支持 fit, transform, inverse_transform
  - 同时支持 element-wise L1 norm 和 Frobenius norm
- 收敛性：包含 simulator 跑出的实验数据并且证明收敛性/准确性（最好有与明文 sklearn 结果的对比，可详见[参考 PR](https://github.com/secretflow/spu/pull/240)）
- 正确性：请确保提交的代码内容为可以直接运行的
- 代码规范：Python 代码需要使用 black+isort 进行格式化（流水线包含代码规范检查卡点）
- 提交说明：关联该 isuue 并提交代码至 https://github.com/secretflow/spu/tree/main/sml
- 特殊说明：若某个特性有特殊的限制，如需要 FM128，需要更多 fxp 等需要在注释文档中明确说明

## 能力要求
- 熟悉经典的机器学习算法
- 熟悉 JAX 或 NumPy，可以使用 NumPy 实现算法

## 操作说明
- 操作指引：https://www.secretflow.org.cn/docs/spu/latest/en-US/getting_started/tutorials/develop_your_first_mpc_application
- 可参考明文实现：[sklearn.decomposition](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.NMF.html#sklearn.decomposition.NMF)
-  [范例](https://github.com/secretflow/spu/blob/main/sml/linear_model/simple_sgd.py)
-  [参考 PR](https://github.com/secretflow/spu/pull/240)

## 开发须知
以下部分代码请必须增加代码注释，对对应代码模块进行说明，包括：
-  __init__函数的超参数含义
-  fit 的具体算法实现说明

### PR DESCRIPTION
**Pull request by tarantula-leo**
# Pull Request

## What problem does this PR solve?
使用 SPU 实现非负矩阵分解算法(NMF) 
Issue Number: Fixed #260

## Possible side effects?

- Performance:
可支持：
1. random init
2. ‘mu’ solver
3. L1/L2正则化
4. frobenius beta loss
5. reconstruction error

- Backward compatibility:

### CODE REVIEW

File: sml/decomposition/nmf.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        ----------
+        X : {array-like}, shape (n_samples, n_features)
+            Training data.
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        Notes
+        -----
+        (1) To prevent overflow error, we can modify the definition of simulator as follows:
```
Comment (deadlywing) [inline sml/decomposition/nmf.py:154]: 我简单测试了一下，FM128似乎不是必须的，使用FM64也不会造成太离谱的错误（随数据量变大误差变大）

File: sml/decomposition/nmf.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        Use 'alpha_W' and 'alpha_H' instead.
+        (3) It's not very convenient to use jax_control_flow in spu right now, so we use 'max_iter' to control
+        the iteration stop, and you can use 'reconstruction_err_' to learn about Frobenius norm of the matrix difference.
+        If you want to get closer to the sklearn, need to keep the number of iterations consistent.
+        (4) Only support 'random' init and 'mu' solver now.
+        """
+        self.fit_transform(X)
+        return self
+
+    def transform(self, X, transform_iter):
```
Comment (deadlywing) [inline sml/decomposition/nmf.py:171]: 非必要不建议改变使用方式，可以增加默认值`transform_iter=None`，然后默认会使用`self._max_iter`

File: sml/decomposition/nmf.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            X, self._l1_ratio, self._alpha_W, self._alpha_H
+        )
+        H_sum, HHt, XHt = None, None, None
+
+        # use multiplicative update solver
+        for _ in range(self._max_iter):
+            W, H_sum, HHt, XHt = update_w(X, W, H, H_sum, HHt, XHt, l1_reg_W, l2_reg_W)
+            if self._update_H:
+                H = update_h(X, W, H, l1_reg_H, l2_reg_H)
+            H_sum, HHt, XHt = None, None, None
```
Comment (deadlywing) [inline sml/decomposition/nmf.py:210]: 这些应该只在H更新的时候才要重新计算吧，如果`self._update_H=False`实际上不用重复计算

File: sml/decomposition/nmf.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+        # compute the regularization parameters
+        l1_reg_W, l1_reg_H, l2_reg_W, l2_reg_H = compute_regularization(
+            X, self._l1_ratio, self._alpha_W, self._alpha_H
+        )
+        H_sum, HHt, XHt = None, None, None
+
+        # use multiplicative update solver
+        for _ in range(self._max_iter):
+            W, H_sum, HHt, XHt = update_w(X, W, H, H_sum, HHt, XHt, l1_reg_W, l2_reg_W)
```
Comment (deadlywing) [inline sml/decomposition/nmf.py:207]: 事实上，可以重新写一下update_w函数，区分是否需要重新计算 H_sum, HHt, XHt

File: sml/decomposition/nmf.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import jax.numpy as jnp
+
+
+def update_w(X, W, H, H_sum, HHt, XHt, l1_reg_W, l2_reg_W):
```
Comment (deadlywing) [inline sml/decomposition/nmf.py:19]: 1. 在当前的beta_loss下，H_sum是不需要的；
2. 需要区分是否要重新计算HHt, XHt

File: sml/decomposition/nmf.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        for _ in range(self._max_iter):
+            W, H_sum, HHt, XHt = update_w(X, W, H, H_sum, HHt, XHt, l1_reg_W, l2_reg_W)
+            if self._update_H:
+                H = update_h(X, W, H, l1_reg_H, l2_reg_H)
+            H_sum, HHt, XHt = None, None, None
+
+        # compute the reconstruction error
+        if self._update_H:
+            self._components = H
+        self.reconstruction_err_ = _beta_divergence(
```
Comment (deadlywing) [inline sml/decomposition/nmf.py:215]: 这个应该也只有在fit和fit_transform的时候计算，也可以一起挪到上面的判断分支里

File: sml/decomposition/tests/nmf_test.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        # Create a simple dataset and random_matrix
+        X = np.random.randint(1, 100, (1000, 10))
+        X = np.array(X, dtype=float)
+        n_samples, n_features = X.shape
+        n_components = 5
+        random_seed = 0
+        random_state = np.random.RandomState(random_seed)
+        A = random_state.standard_normal(size=(n_components, n_features))
+        B = random_state.standard_normal(size=(n_samples, n_components))
+        l1_ratio = 0
```
Comment (deadlywing) [inline sml/decomposition/tests/nmf_test.py:79]: 测试的时候，建议也把 l1_ratio设置为非0吧

File: sml/decomposition/tests/nmf_test.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+import spu.utils.simulation as spsim
+
+# Add the sml directory to the path
+sys.path.append(os.path.join(os.path.dirname(__file__), '../../../'))
+
+from sml.decomposition.nmf import NMF
+
+
+class UnitTests(unittest.TestCase):
+    def test_nmf(self):
```
Comment (deadlywing) [inline sml/decomposition/tests/nmf_test.py:31]: 建议把测试分开吧，一个测试建议只测试一个功能～

File: sml/decomposition/nmf.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+            X, W, H, self._beta_loss, square_root=True
+        )
+
+        return W
+
+    def inverse_transform(self, X):
+        return jnp.dot(X, self._components)
```
Comment (deadlywing) [inline sml/decomposition/nmf.py:222]: 需要判断是否已经fit

File: sml/decomposition/tests/nmf_test.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        random_state = np.random.RandomState(random_seed)
+        A = random_state.standard_normal(size=(n_components, n_features))
+        B = random_state.standard_normal(size=(n_samples, n_components))
+        l1_ratio = 0
+        alpha_W = 0.01
+
+        # Run the simulation
+        W, H, X_reconstructed = spsim.sim_jax(sim, proc1)(X, A, B)
+        print("W_matrix_spu: ", W)
+        print("H_matrix_spu: ", H)
```
Comment (deadlywing) [inline sml/decomposition/tests/nmf_test.py:85]: print到控制台的内容过于多了，如果可以的话，建议用一些assert语句来替代；
如np.testing.asseer*中的api，例如可以测试spu和sklearn的 W dot H 是否接近（可以选择合适的tol）；

Comment (deadlywing) [reply sml/decomposition/tests/nmf_test.py:85]: 另外，也可以测试一下reconstruction_err_的值哈

File: sml/decomposition/emulations/nmf_emul.py
Comment (deadlywing) [inline sml/decomposition/emulations/nmf_emul.py:1]: emul可以参考test文件进行对称的修改；

**Comment by deadlywing** [comment]
PS：之前不能在同一个emul文件中进行多次emulation的bug找到了，是grpc的问题，只要把grpc升级到最新版本，然后清理一下缓存(bazel clean)然后重新运行即可

File: sml/decomposition/nmf.py (hunk 1 above)
Comment (tarantula-leo) [reply sml/decomposition/nmf.py:154]: 之前测过一组数据会存在一定的问题 应该是数据量变大了 建议还是使用128

File: sml/decomposition/nmf.py (hunk 2 above)
Comment (tarantula-leo) [reply sml/decomposition/nmf.py:171]: 是为了保持迭代次数和sklearn一致 self._max_iter是fit时候使用的 transform时不同 需要传参进来 sklearn里用条件判断是否停止迭代 不需要调整max_iter

File: sml/decomposition/tests/nmf_test.py (hunk 10 above)
Comment (tarantula-leo) [reply sml/decomposition/tests/nmf_test.py:85]: 1. 哪些数据需要Print 哪些使用assert？
2. reconstruction_err_之前沟通是不需要测试输出 用户有需要自行调用查看

File: sml/decomposition/nmf.py (hunk 2 above)
Comment (deadlywing) [reply sml/decomposition/nmf.py:171]: ok，我的意思只是希望即使不加transform_iter也能正常运行，加一个默认值而已

File: sml/decomposition/nmf.py (hunk 1 above)
Comment (deadlywing) [reply sml/decomposition/nmf.py:154]: 很多涉及矩阵乘的算法，数据量大的时候都得128，这是普遍现象；
我感觉这里的描述有点64就完全不行的意思，可以适当修改一下描述。

主要单元测试里我感觉64就差不多够了，不然测试的运行时间太久也不太好～

（PLUS，我看了一下sklearn的单元测试，其实用的数据集大小也不是很大...-.-）

File: sml/decomposition/tests/nmf_test.py (hunk 10 above)
Comment (deadlywing) [reply sml/decomposition/tests/nmf_test.py:85]: 1. 这个主要还是需要你测试一下，能assert尽量assert；如果需要print尽量也限制大小，比如只打印几行几列
2. 单元测试是为了验证正确性，，能覆盖的尽量覆盖.（我之前的意思是不需要直接在fit里return出来）

File: sml/decomposition/nmf.py (hunk 1 above)
Comment (tarantula-leo) [reply sml/decomposition/nmf.py:154]: 在Notes里说明下大数据时情况 测试单元里也使用64 emulation中也调整成64 fxp_fraction_bits都保持30 这样调整？

Comment (deadlywing) [reply sml/decomposition/nmf.py:154]: 如果用FM64，fxp_fraction_bits就不用调整了，默认值就好

File: sml/decomposition/nmf.py (hunk 5 above)
Comment (tarantula-leo) [reply sml/decomposition/nmf.py:19]: 1. H_sum留口子给后续补充'kullback-leibler'模式用的 注释说明下
2. 是否要重新计算HHt, XHt指的是什么？在update_W函数里使用update_H判断是否需要计算HHt, XHt？

Comment (deadlywing) [reply sml/decomposition/nmf.py:19]: 1. ok
2. 是的，如果H不变的话，那俩是不用一直重复计算的

File: sml/decomposition/tests/nmf_test.py (hunk 10 above)
Comment (tarantula-leo) [reply sml/decomposition/tests/nmf_test.py:85]: 之前考虑不使用assert的原因是 在先fit再transform的seperate场景中 transform的迭代次数很难和sklearn的保持一致 assertinverse_transform结果时会出错 而且基本第一遍跑都会报错 又考虑seperate和fit_transform模式二者在输出上保持统一 就都没有使用assert了 建议还是使用print 可以只打几行

Comment (tarantula-leo) [reply sml/decomposition/tests/nmf_test.py:85]: sklearn transform时的model.n_iter_次数是不记录的 只会记录fit/fit_transform里的迭代次数

Comment (deadlywing) [reply sml/decomposition/tests/nmf_test.py:85]: 如果迭代次数不一样，最后结果会差很多么？调整tol不能通过测试么

Comment (tarantula-leo) [reply sml/decomposition/tests/nmf_test.py:85]: 在相同迭代次数的情况下 如果是用FM64 tol设置为1e-1 也会通过不了测试 迭代次数不同肯定是不行了

Comment (tarantula-leo) [reply sml/decomposition/tests/nmf_test.py:85]: fit/fit_transform里是因为使用sklearn会达到迭代次数上限 所以可以保持一致

Comment (deadlywing) [reply sml/decomposition/tests/nmf_test.py:85]: > 在相同迭代次数的情况下 如果是用FM64 tol设置为1e-1 也会通过不了测试 迭代次数不同肯定是不行了

有个问题，那如果FM128，fxp=30，迭代次数不同能通过测试么

Comment (tarantula-leo) [reply sml/decomposition/tests/nmf_test.py:85]: 不可以 FM128，fxp30， tol设置为1e-1 迭代次数保持一致 可以通过

Comment (tarantula-leo) [reply sml/decomposition/tests/nmf_test.py:85]: 建议使用128/30的参数 否则不是seperate模式也会出现assert不通过

Comment (deadlywing) [reply sml/decomposition/tests/nmf_test.py:85]: 能用64尽量用64，不行就128；
fit_transform用64能行不

Comment (deadlywing) [reply sml/decomposition/tests/nmf_test.py:85]: > 建议使用128/30的参数 否则不是seperate模式也会出现assert不通过

OK，那就都用128吧；

Comment (tarantula-leo) [reply sml/decomposition/tests/nmf_test.py:85]: tol设置为0.1的话是不行的

Comment (tarantula-leo) [reply sml/decomposition/tests/nmf_test.py:85]: 试了下 inverse的矩阵在普通模式下assert时也会偶现不通过 W、H矩阵都可以通过 是不是考虑inverse这部分也通过print？

File: sml/decomposition/nmf.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import jax.numpy as jnp
+
+
+def update_w(X, W, H, H_sum, HHt, XHt, l1_reg_W, l2_reg_W, update_H=True):
```
Comment (deadlywing) [inline sml/decomposition/nmf.py:19]: update_H 这个参数没有用到

File: sml/decomposition/emulations/nmf_emul.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+        H_Sklearn = model.components_
+        X_reconstructed_Sklearn = model.inverse_transform(W_Sklearn)
+        err = model.reconstruction_err_
+        print("reconstruction_error_sklearn: ", err)
+        assert np.allclose(W_Sklearn, W, atol=5e-1)
+        assert np.allclose(H_Sklearn, H, atol=5e-1)
+        assert np.allclose(X_reconstructed_Sklearn, X_reconstructed, atol=5e-1)
+
+        # sklearn_seperate
+        model = SklearnNMF(
```
Comment (deadlywing) [inline sml/decomposition/emulations/nmf_emul.py:88]: 这个函数里不需要跑sklearn的seperate吧

File: sml/decomposition/emulations/nmf_emul.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+        W = model.fit_transform(X)
+        H = model._components
+        X_reconstructed = model.inverse_transform(W)
+        err = model.reconstruction_err_
+        return W, H, X_reconstructed, err
+
+    try:
+        # bandwidth and latency only work for docker mode
+        conf_path = "sml/decomposition/emulations/3pc.json"
+        emulator = emulation.Emulator(conf_path, mode, bandwidth=300, latency=20)
```
[the remark this answers is not in the document]
Comment (tarantula-leo) [reply sml/decomposition/emulations/nmf_emul.py:47]: up和down要放到try/finally里么？

[the remark this answers is not in the document]
Comment (deadlywing) [reply sml/decomposition/emulations/nmf_emul.py:47]: 可以在main函数里整体try catch，emulator一次up/down即可

[the remark this answers is not in the document]
Comment (tarantula-leo) [reply sml/decomposition/emulations/nmf_emul.py]: Recheck.

**Review by deadlywing — APPROVED** [review]
LGTM
